wasm-rpc/src/lib.rs

```diff
@@ -304,6 +402,576 @@ fn build_tree(node: &WitNode, nodes: &[WitNode]) -> Value {
     }
 }
 
+// An efficient read representation of a value with type information at every node
+// Note that we do an extra annotation of AnalysedType at complex structures to fetch their type information with 0(1)
+// However the typical use of TypeAnnotatedValue is similar to using a Json/Yaml etc
+#[derive(Clone, Debug)]
+pub enum TypeAnnotatedValue {
+    Bool(bool),
+    S8(i8),
+    U8(u8),
+    S16(i16),
+    U16(u16),
+    S32(i32),
+    U32(u32),
+    S64(i64),
+    U64(u64),
+    F32(f32),
+    F64(f64),
+    Chr(char),
+    Str(String),
+    List(ListValue),
+    Tuple(TupleValue),
+    Record(RecordValue),
+    Flags(FlagValue),
+    Variant(VariantValue),
+    Enum(EnumValue),
+    Option(OptionValue),
+    Result(ResultValue),
+    Handle(ResourceValue),
+}
+
+#[derive(Clone, Debug)]
+pub struct EnumValue {
+    typ: Vec<String>,
+    value: String,
+}
+
+#[derive(Clone, Debug)]
+pub struct OptionValue {
+    typ: AnalysedType,
+    value: Option<Box<TypeAnnotatedValue>>,
+}
+
+#[derive(Clone, Debug)]
+pub struct FlagValue {
+    typ: Vec<String>,
+    value: Vec<String>, // value should be a subset of typ field here.
+}
+
+#[derive(Clone, Debug)]
+pub struct VariantValue {
+    typ: Vec<(String, Option<AnalysedType>)>,
+    case_name: String,
+    case_value: Option<Box<TypeAnnotatedValue>>,
+}
+
+#[derive(Clone, Debug)]
+pub struct TupleValue {
+    typ: Vec<AnalysedType>,
+    value: Vec<TypeAnnotatedValue>,
+}
+
+#[derive(Clone, Debug)]
+pub struct RecordValue {
+    typ: Vec<(String, AnalysedType)>,
+    value: Vec<(String, TypeAnnotatedValue)>,
+}
+
+//
+// The law here is:
+//     let mut types = Vec::new();
+//       for value in values {
+//          types.push(value.analysed_typ());
+//       }
+//       let head = types.map(|value| value.into())).head
+//       types.forall(types == head)
+//
+#[derive(Clone, Debug)]
+pub struct ListValue {
+    typ: AnalysedType,
+    values: Vec<TypeAnnotatedValue>,
+}
+
+#[derive(Clone, Debug)]
+pub struct ResultValue {
+    ok: Option<Box<AnalysedType>>,
+    error: Option<Box<AnalysedType>>,
+    value: Result<Option<Box<TypeAnnotatedValue>>, Option<Box<TypeAnnotatedValue>>>,
+}
+
+#[derive(Clone, Debug)]
+pub struct ResourceValue {
+    id: AnalysedResourceId,
+    resource_mode: AnalysedResourceMode,
+    uri: Uri,
+    resource_id: u64,
+}
+
+impl TypeAnnotatedValue {
+    pub fn analysed_typ(&self) -> AnalysedType {
```

Comment:
This should be a ``From<TypedAnnotatedValue> for AnalysedType` instance instead

Reply:
Okay

Reply:
Done